python/cuml/neighbors/kneighbors_classifier.pyx

```diff
@@ -0,0 +1,257 @@
+        :param X:
+        :param y:
+        :param convert_dtype:
+        :return:
+        """
+        super(KNeighborsClassifier, self).fit(X, convert_dtype)
+
+        self.y, _, _, _, _ = \
+            input_to_dev_array(y, order='F', check_dtype=np.int32,
+                               convert_to_dtype=(np.int32
+                                                 if convert_dtype
+                                                 else None))
+
+        self.handle.sync()
+
+    def predict(self, X, convert_dtype=True):
+        """
+        Use the trained k-nearest neighbors classifier to
+        predict the labels for X
+        :param X:
+        :param convert_dtype:
+        :return:
+        """
+        return self._classify(X, convert_dtype)
+
+    def _classify(self, X, convert_dtype=True, _knn_indices=None):
+
+        knn_indices = self.kneighbors(X, return_distance=False,
+                                      convert_dtype=convert_dtype) if \
+            _knn_indices is None else _knn_indices
+
+        cdef uintptr_t inds_ctype
+        inds, inds_ctype, n_rows, _, _ = \
+            input_to_dev_array(knn_indices, order='C', check_dtype=np.int64,
+                               convert_to_dtype=(np.int64
+                                                 if convert_dtype
+                                                 else None))
+
+        out_cols = self.y.shape[1] if len(self.y.shape) == 2 else 1
+
+        out_shape = (n_rows, out_cols) if out_cols > 1 else n_rows
+
+        classes = rmm.to_device(zeros(out_shape,
+                                      dtype=np.int32,
+                                      order="C"))
+
+        cdef vector[int*] *y_vec = new vector[int*]()
+
+        # If necessary, separate columns of y to support multilabel
+        # classification
+        cdef uintptr_t y_ptr
+
+        for i in range(out_cols):
+            col = self.y[:, i] if out_cols > 1 else self.y
+            y_ptr = get_dev_array_ptr(col)
+            y_vec.push_back(<int*>y_ptr)
+
+        cdef uintptr_t classes_ptr = get_dev_array_ptr(classes)
+
+        cdef cumlHandle* handle_ = <cumlHandle*><size_t>self.handle.getHandle()
+
+        knn_classify(
+            handle_[0],
+            <int*> classes_ptr,
+            <int64_t*>inds_ctype,
+            deref(y_vec),
+            <size_t>X.shape[0],
+            <int>self.n_neighbors
+        )
+
+        self.handle.sync()
+        if isinstance(X, np.ndarray):
+            return np.array(classes)
+        elif isinstance(X, cudf.DataFrame):
+            return cudf.DataFrame.from_gpu_matrix(X)
+        else:
+            return classes
+
+    def predict_proba(self, X, convert_dtype=True):
+        """
+        Use the trained k-nearest neighbors classifier to
+        predict the label probabilities for X
+        :param X:
+        :param convert_dtype:
+        :return:
+        """
+        return self._predict_proba(X, convert_dtype)
```

Comment:
same question as in `predict`, why create the private `_predict_proba` since it takes the same parameters and seems to only be called here?